### DISCUSSION TITLE
`mikro-orm schema:update -r` fails to create a schema when it already exists

### DISCUSSION THREAD

**Discussion by tukusejssirs** [mikro-orm/mikro-orm#2588] [Q&A]
It fails with `DriverException: create schema "schema"; - schema "schema" already exists` error.

When I run `mikro-orm schema:update -d`, it outputs `create schema "schema";`.

Expected command: `create schema if not exists "schema";`.

**Comment by B4nan** [comment]
v4 or v5? only place where this could be triggered is [here](https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/schema/SchemaGenerator.ts#L158), and that would mean the diffing did not see your existing schema - which could point to a misconfigured `search_path` most probably (or some other configuration issue, e.g. do you specify `schema` option in the ORM config?). 

if you can create a full reproduction, will be happy to look into it.

**Reply by tukusejssirs** [reply]
> v4 or v5?

I only use v5.

> which could point to a misconfigured `search_path` most probably

Where can I configure `search_path`?

I use multiple/many schemas. Some entities use only a single schema, others use schemas configured at runtime (and I use `schema: '*'` in the `@Entity()` decorator).

> or some other configuration issue, e.g. do you specify `schema` option in the ORM config

I have no `schema` option configured in `mikro-orm.config.ts`. I have configured the schema names on in `@Entity()` and `withSchema()` when building the queries.

Note that this error occured after I defined a new entity with a new (not used until now) schema.

**Reply by B4nan** [reply]
https://www.postgresql.org/docs/9.6/ddl-schemas.html#DDL-SCHEMAS-PATH

**Reply by tukusejssirs** [reply]
> https://www.postgresql.org/docs/9.6/ddl-schemas.html#DDL-SCHEMAS-PATH

I’ve read this already.

---

Here I describe how we structured our DB and how we use it:

1. We operate with database via API (backend) only, therefore there are only pre-defined database operations that an authenticated user (frontend) can do.
2. Only a single DB user has access the database and that one is used only by MikroORM (or from `psql` for debugging purposes).
2. Database tables are divided into the following schemas:
   - `md`:
      - general metadata, esp about machines;
   - `mx` (where `x` is in unique integer assigned to a machine):
      - contains all tables required for monitoring a particular machine;
      - more or less, each machine uses the same group of tables, but not all machines contains all the tables;
   - `user`:
      - user and auth related tables.

Basically, I could prepend or append the schema name to the table names and use a single schema in the database, but from my point of view it is much easier to work with schemas, as for example I can easily delete all machine-related tables by running `drop schema mx cascade;` without worrying if the machine use `mx.table_name` or not. Also I think that moving schema names to tables name would not be very DRY (although that is not a big deal in this particular case).

Until now (prior to using ORM or Node.js), I happily used qualified names without any issue. Now I want to use MikroORM and all its features to simplify the DB management and I don’t think that using a single schema would solve my issues.

As for `SET search_path TO myschema,public;`, I know of this command, but I actually don’t use it. I _always_ (even in `psql`) use qualified table names (like `select * from schema.table;`).

However, I’d be very grateful if you point out what is bad about this particular DB design.

---

Back to the original issue of this GitHub Discussion: when I rung `mikro-orm schema:create -r`, all I want is to:
1. create any schema/table/column/whatever _if it does not already exist_;
2. update any table/column/whatever _if it was changed_.

Can MikroORM be configured like that?

**Reply by tukusejssirs** [reply]
Another related issue: the generated command to rename an index does not include a schema name before the index name, although the table uses one (as all tables in our database), for example:

```sql
-- Current
alter index "table_id_idx" rename to "name_id_index";

-- Expected
alter index "md"."table_id_idx" rename to "name_id_index";
```

**Reply by tukusejssirs** [reply]
Another related issue (related to #507): `schema:update` always has the same updates. In this parcular entity, there is no column set to be `nullable` and not one of them has default value set to `null`.

```ts
// Entity
import {Entity, Index, PrimaryKey, Property} from '@mikro-orm/core'

@Entity({schema: 'md', tableName: 'table'})
@Index({properties: ['id'], type: 'btree'})
export class Ent {
  @PrimaryKey()
  id!: number

  @Property({columnType: 'VARCHAR', unique: true, default: ''})
  name: string

  @Property({columnType: 'TIMESTAMP WITH TIME ZONE', fieldName: 'created', defaultRaw: 'NOW()'})
  createdAt = new Date()

  @Property({columnType: 'TIMESTAMP WITH TIME ZONE', fieldName: 'last_modified', defaultRaw: 'NOW()', onUpdate: () => new Date()})
  updatedAt = new Date()
}
```

`schema:update` always wants to run the following commands:

```sql
alter table "md"."table" drop constraint if exists "md.table_created_check";
alter table "md"."table" alter column "created" type TIMESTAMP WITH TIME ZONE using ("created"::TIMESTAMP WITH TIME ZONE);
alter table "md"."table" drop constraint if exists "md.table_last_modified_check";
alter table "md"."table" alter column "last_modified" type TIMESTAMP WITH TIME ZONE using ("last_modified"::TIMESTAMP WITH TIME ZONE);
```

**Update:** Even if replacing `default: ''` with `nullable: true` on `name` property (the former was obviously a mistake), the same commands are generated.

**Reply by B4nan** [reply]
You are missing my point. This is not about using `if not exists` clause. The ORM does not see your existing schema, that is what you need to resolve. And that is a permission issue, nothing inside ORM config. Of course I can just use the `if not exists` clause, but schema diffing would still not work. You would get pass this one query that tries to create the schema, but you would end up with not working schema diffing for those schemas that are not visible by the ORM. You would never get anything more than a `create schema if not exists` + `create table if not exists`. It could never do alter table queries, as it can't see your schema at all, hence it can't see the tables inside your schema.

> Another related issue: the generated command to rename an index does not include a schema name before the index name, although the table uses one (as all tables in our database), for example:

This seems valid, will take a look.

> Another related issue (related to #507): schema:update always has the same updates. In this parcular entity, there is no column set to be nullable and not one of them has default value set to null.

Instead of `TIMESTAMP WITH TIME ZONE` it should be just `timestamptz` and you need to privide the length, as the ORM defaults to 0 and postgres defaults to 6. 

```ts
  @Property({columnType: 'TIMESTAMPTZ(0)', fieldName: 'created', defaultRaw: 'now()'})
  createdAt = new Date();
```

This is definitly weird, will see if I can do something about the length parameter default value. But you don't really need that, if you would specify the property like this, it would be all inferred:

```ts
  @Property({fieldName: 'created', defaultRaw: 'now()'})
  createdAt: Date = new Date(); // you need to provide the type explicitly for reflect-metadata to work
```

Or you can use one of the mapped types:

```ts
import { t } from '@mikro-orm/core';

  @Property({type: t.datetime, fieldName: 'created', defaultRaw: 'now()'})
  createdAt = new Date(); // you need to provide the type explicitly for reflect-metadata to work
```

Both of this will use `timestamptz` aka `timestamp with time zone` in postgres. It is the default column type for date columns.

**Reply by tukusejssirs** [reply]
> for those schemas that are not visible by the ORM

Well, I still might not fully understand what you’re trying to tell me, however, _all_ schemas are created by MikroORM only. Actually the entire DB management is done via MikroORM. Is it still an issue then?

> It could never do alter table queries, as it can't see your schema at all, hence it can't see the tables inside your schema.

Why if in this particular case I created an entity with `@Entity({schema: 'md', tableName: 'table'})`?

> > Another related issue: the generated command to rename an index does not include a schema name before the index name, although the table uses one (as all tables in our database), for example:
>
> This seems valid, will take a look.

Thanks!

> Instead of `TIMESTAMP WITH TIME ZONE` it should be just `timestamptz` and you need to privide the length, as the ORM defaults to 0 and postgres defaults to 6.

Thanks for the tip. Now it works as expected (I used `timestamptz(3)` though).

**Reply by B4nan** [reply]
> Well, I still might not fully understand what you’re trying to tell me, however, all schemas are created by MikroORM only. Actually the entire DB management is done via MikroORM. Is it still an issue then?

MikroORM does not handle permissions in any way, and that is most probably your problem. So once again, from inside the ORM, your schema does not exist, that is the reason why it tries to create it. You keep asking for a different query, but the real solution is to let the ORM know the schema exists, so there will be no "create schema" query at all.

> Why if in this particular case I created an entity with @Entity({schema: 'md', tableName: 'table'})?

This has nothing to do with entity definition, again, this is permission issue.

This is the query that is used for retrieving all existing tables, from all schemas (that are visible to the current user - that is the part where it probably fails for you) exlucing only internal ones (prefixed with `pg_`). 

```sql
select table_name, nullif(table_schema, 'public') as schema_name, (select pg_catalog.obj_description(c.oid) from pg_catalog.pg_class c where c.oid = (select ('"' || table_schema || '"."' || table_name || '"')::regclass::oid) and c.relname = table_name) as table_comment 
from information_schema.tables 
where table_schema not like 'pg_%' and table_schema != 'information_schema' and table_name != 'geometry_columns' and table_name != 'spatial_ref_sys' and table_type != 'VIEW' 
order by table_name
```

It is probably not returning tables from the schema. 

One problem I could imagine is that your schema is empty, in such case it would probably end up with this problem too. Can you verify that? I guess we should use the `if not exists` anyway, to mitigate this particular issue. But if you know there are tables in that schema, what I am talking about above is still valid, it won't help much, as without the ORM seeing your tables, there can't be any diffing on them.

**Reply by tukusejssirs** [reply]
> MikroORM does not handle permissions in any way, and that is most probably your problem. So once again, from inside the ORM, your schema does not exist, that is the reason why it tries to create it. You keep asking for a different query, but the real solution is to let the ORM know the schema exists, so there will be no "create schema" query at all.

Okay, I’ve tested this using `postgres` (superadmin) user, created the DB structure using MikroORM as `postgres` user (therefore that user should have all permissions over everything in the DB), but the issue persists.

Or are there any other permissions that are not (at least implicitly) granted to `postgres` user?

> This is the query that is used for retrieving all existing tables, from all schemas (that are visible to the current user - that is the part where it probably fails for you) exlucing only internal ones (prefixed with `pg_`).

When I login using `psql $db_name` as `postgres` user in terminal and run that SQL command, it lists all the tables from all schemas without any issues.

> One problem I could imagine is that your schema is empty

No schema is empty. Every schema contains at least one table.

**Reply by B4nan** [reply]
> When I login using psql $db_name as postgres user in terminal and run that SQL command, it lists all the tables from all schemas without any issues.

But apparently when its executed from the ORM, it does not. If it happens locally, try putting some console logs to the code where this is executed to double check.

https://github.com/mikro-orm/mikro-orm/blob/master/packages/knex/src/schema/DatabaseSchema.ts#L53

I guess you do have debug mode enabled, right? You should see the connection string being used from inside the ORM, to validate the same user is being used. In latest v5 the queries are not being logged by default from inside CLI commands, you need to enable `MIKRO_ORM_VERBOSE` env var:

```sh
$ MIKRO_ORM_VERBOSE=1 mikro-orm schema:update -d
```

As asked before, if you believe this is not about persmissions and you are able to create a full reproduction (so complete but minimal project, not just code snippets - including a docker compose script ideally, as this can be as well about configuration of the database or the ORM), I will be happy to look into it. But what you gave me is already present in the test suite, and works fine.

**Reply by tukusejssirs** [reply]
> But apparently when its executed from the ORM, it does not.

Right.

> I guess you do have debug mode enabled, right?

Yeah, but not by default. Unfortunately, when the issue occured, I didn’t check the debugs messages (appart from `schema:update -d`) and for now I resolved the issue manually, therefore my immediate issue is solved _for now_.

> if you believe this is not about persmissions and you are able to create a full reproduction

Yes, I believe this is not about permissions, but unfortunately ATM I am unable to crate a full repro. When the issue shows up again, I’ll try to debug it more in depth and hopefully I’ll be able to create a full repro.

---

Would be feasable to make `--schema` option accept an array of schema names (e.g. using `--schema one,two,three`)?
